Readme.md
# Dropper

> __Dropper__ is a ( nodeJS ) __filter stream__ that produces __fixed size__ data packets from __any stream__, mantaining __very good perfomances__ .

> __Dropper__ inherits from __Stream__, then you can use it in the way you __already__ use other streams
> ( _pausing_, _resuming_, _writing_.. ) .

> It simply _"divides"_ or _"bufferizes"_ incoming data packets, depending on whether 
> the received packet is smaller or larger than its own capacity.

---------

###Installation

> **Current __Stable__ Version: 0.0.7 , compatible with nodeJS >= v0.4.x**

> with __npm__ :

``` bash
 $ npm install dropper
```

> or clone with __git__ :


``` bash
 $ git clone git://github.com/rootslab/dropper.git
```
---------

###Simple Usage

``` javascript
  var opt = { pad : 0x0 }, // <- add leading 0s to last packet for mantaining a constant size 
      dropSize = 64 * 1024, // <- default bytes value
      filter = new dropper( dropSize, opt );
```

> Try this, or just run the __[example](https://github.com/rootslab/dropper/blob/master/examples/)__ :

``` javascript
  var log = console.log,
      Dropper = require( 'dropper' ),
      path = __filename, // <- a test file path
      filter = new Dropper( 36 ); // output -> 36 bytes data packets

 /*
  * all data packets will have fixed length ( dropSize ),
  * except the last that could be <= dropSize
  */  
  filter.on( 'data', function ( data ) {
      log( 'data:', data );   // <- print data ( do something with data )
  } );
  
  /*
  * stream output -> 60 bytes packets, it's the max limit,
  * packets are not guaranteed to have fixed size.
  */
  stream = fs.createReadStream( path, { bufferSize : 60 } );
  ..
  
 /*
  *  nodeJS versions < 0.5 don't support pipe chaining,
  *  because don't return the destination stream.
  *  With dropper pipe(), instead, is always possible.
  */
  stream.pipe( filter );
  filter.pipe( fs.createWriteStream( __dirname + '/drops.txt' ) );
  ..
  
```
----------
###Run Tests & Benchmarks

> Just :


``` bash
 $ node test/data_integrity.js
 $ node test/benchmark.js [ 1 ] // <- a truthy value for switching from dropper to block-stream module use
```

## License 

(The MIT License)

Copyright (c) 2011 Guglielmo Ferri &lt;44gatti@gmail.com&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md
0.0.7 / 2010-10-20
==================
  
  * added benchmarks for dropper and block-stream

0.0.6 / 2010-10-20
==================
  
  * added padding choice for the last packet

0.0.5 / 2010-10-18
==================
  
  * better performances

0.0.4 / 2010-10-18
==================
  
  * less Buffer instances

0.0.3 / 2010-10-18
==================
  
  * went back to code version 0.0.1
  * added Buffer instances and slice()
  * thanks @isaacs

0.0.2 / 2010-10-17
==================
  
  * resolved bug when input data length < output data length ( dropSize )
  * resolved bug when input data length > output data length
  * added squueze and chargeAndSqueeze methods
  
0.0.1 / 2010-10-15
==================

  * Initial release
  * bugged version, drop data is obviously overwritten if aren't copied quickly


test/benchmark.js
var log = console.log,
    assert = require( 'assert' ),
    crypto = require( 'crypto' ),
    os = require( 'os' ),
    fs = require( 'fs' ),
    // switch block-stream or dropper
    useBlockStream = ( process.argv.length > 2 ) ? ( ( process.argv[ 2 ] & 1 ) ? true : false ) : false,
    Dropper = require( ( useBlockStream ) ? 'block-stream' : '../index' ),
    // colors
    rc = '', gc = '', yc = '', tc = '', bc = '', xc = '', ec = '', wc = '',
    color = ( os.type() === 'Linux' ) ? (
      ( rc = '\033[1;31m' ) && ( gc = '\033[1;32m' ) && ( yc = '\033[1;33m' ) &&
        ( tc = '\033[1;44m' ) && ( bc = '\033[1;39m' ) && ( xc = '\033[1;35m' ) &&
          ( wc = '\033[1;42m') && ( ec = '\033[0m' ) ) : null,
    // vars
    isize = 0, tsize = 0, fsize = 0, psize = 0, cycles = 0, 
    wtime = 0, ftime = 0, stime = 0, expected = 0,
    filter = null, paper = null, stack = null,
    rstart = null, rend = null, pstart = null, pend = null,
    hash = {},
    // methods
    recTime = function () {
        var me = this,
            stream = hash[ me.info ],
            elapsed = 0;
            
        if ( stream ) { 
            stream.time.end = Date.now();
            stream.packets.counter += 1;
            if( stream.packets.counter >= stream.packets.expected ) {
                elapsed = stream.time.end - stream.time.start;
                log( tc + me.info + ' stream ->', ec, '\n' + bc, stream  );
                ( stream.packets.counter > stream.packets.expected ) ? 
                    log( rc + 'Damn! packets received are more than expected!' + ec ) : null;
                log( wc + 'elapsed time:', elapsed, 'ms', ec, '\n' );
            }
        } else {
            stream = hash[ me.info ] = {
                packets : {
                    expected : me.expected,
                    counter : 0
                },
                time : {
                    start : Date.now(),
                    end : null
                }
            };
        }
    },
    writeData = function ( stream, howmanytimes, packetsize, totalsize ) {
        var i = 0,
            j = 0,
            expected = Math.ceil( howmanytimes * totalsize / packetsize );
        recTime.bind( { info : 'source', expected : expected } )();        
        for ( ; i < howmanytimes; ++i ) {
            for ( j = packetsize; j <= totalsize; j += packetsize ) {
                stream.write( stack.slice( j - packetsize, j ) );
                recTime.bind( { info : 'source', expected : expected } )()
            }
        }
        stream.end();
    },
    go = function ( ) {
        log( yc + 'piping ( bytes ):', isize, '->', fsize, '->', psize, ec, '\n' );
        hash = {};
        filter = new Dropper( fsize, { nopad : true } ); // <- { nopad : true } is only for block-stream lib
        expected = Math.ceil( cycles * tsize / fsize );
        fstart = recTime.bind( {
            info : 'filter',
            expected : expected
        } );
        fend = recTime.bind( {
            info : 'filter'
        } );
        filter.on( 'data', fstart );
        filter.on( 'end', fend );
        
        paper = new Dropper( psize, { nopad : true } ); // <- { nopad : true } is only for block-stream lib
        expected = Math.ceil( cycles * tsize / psize );
        pstart = recTime.bind( {
            info : 'paper',
            expected : expected
        } );
        pend = recTime.bind( { info : 'paper' } );
        paper.on( 'data', pstart );
        paper.on( 'end', pend );
        
        filter.pipe( paper );
        writeData( filter, cycles, isize, tsize );
    },
    init = function () {
        tsize = 6 * 1024 * 1024; // bytes
        stack = new Buffer( tsize );
        log( '\n' + rc, 'Using', ( ( useBlockStream ) ? 'block-stream' : 'dropper' ), 'library..', ec, '\n' );
        log( tc, 'Filling the stack buffer with', ( tsize / 1024 / 1024 ).toFixed( 2 ), 'MBytes of data', ec );
        for ( i = 0; i < tsize; stack[ i++ ] = i & 255 );
    }; 


init();

log( '\n' + gc + '-> starting test 1 <-' + ec + '\n' );
cycles = 1024;
isize = 64 * 1024;
fsize = psize = isize;
go();

log( '\n' + gc + '-> starting test 2 <-' + ec + '\n' );
cycles = 256;
isize = 64 * 1024;
fsize = Math.ceil( isize * 3.141516 );
psize = Math.ceil( isize / 6 );
go();

log( '\n' + gc + '-> starting test 3 <-' + ec + '\n' );
cycles = 256;
isize = 64 * 1024;
fsize = Math.ceil( 1 + isize / 2 );
psize = Math.ceil( isize - 1 );
go();


test/data_integrity.js
var log = console.log,
    assert = require( 'assert' ),
    crypto = require( 'crypto' ),
    fs = require( 'fs' ),
    Dropper = require( '../index' ),
    path =  __filename, //  __dirname + '/../examples/numbers.txt',
    checksum = null,
    output = null, o = 0,
    filter = null, sourceStream = null,
    rc = '\033[1;31m', gc = '\033[1;32m', yc = '\033[1;33m',
    tc = '\033[1;44m', bc = '\033[1;39m', xc = '\033[1;35m',
    ec = '\033[0m',
    delay = 30;

fs.readFile( path, null, function ( err, fdata ) {
    if ( err ) {
        log( 'fs error:', err.stack );
    }
    checksum = crypto.createHash( 'sha1' ).update( fdata ).digest( 'hex' );
    log( '\n' + yc + 'Starting Data Integrity Test..' + ec +'\n' );
    
    output = new Buffer( fdata.length );
    
    // dropper

    filter = new Dropper( 36 );

    filter.on( 'data', function ( data ) {
        filter.pause();  // <- for a better output log
        setTimeout( function (){
            log( tc + 'filter emits data:', data, ec );
            data.copy( output, o );
            o += data.length;
            if ( o === output.length ) {
                // data.length === filter.dropsize, except for last packet ( <= )
                try {
                    assert.ok( data.length <= filter.dropSize, 'Test Failed: data_integrity, last packet bad size ' );
                    log( yc + 'Test ok ( last output packet length <= dropSize ):', gc + data.length, '<=', rc + filter.dropSize + ec );
                } catch ( err ) {
                    log( rc + '\nsomething went wrong..\n' );
                    log( err.stack, ec );
                    filter.destroy();
                }
            } else {
                try {
                    assert.strictEqual( data.length, filter.dropSize, 'Test Failed: data_integrity, bad packet size ' );
                    log( yc + 'Test ( output packet length === dropSize ) ok:', gc + data.length, '===', rc + filter.dropSize + ec );
                } catch ( err ) {
                    log( rc + '\nsomething went wrong..\n' );
                    log( err.stack, ec );
                    filter.destroy();
                }
            }
            filter.resume(); // <- for a better output log
        }, delay );
    } );

    filter.on( 'drain', function () {
    } );

    filter.on( 'end', function () {
        setTimeout( function () {
            log( yc + 'filter stream ends,', filter.sent, 'packet(s) sent,', filter.received, 'received' + ec );
            o = 0;
            var resultChecksum = crypto.createHash( 'sha1' ).update( output ).digest( 'hex' );
            log( xc + 'source file:', path + '\nsize:', fdata.length, 'bytes' + ec );
            try {
                assert.strictEqual( resultChecksum, checksum, 'Test Failed: data_integrity, bad checksum ' );
                log( yc + 'Test ok:', gc + checksum, '===', rc + resultChecksum + ec );
            } catch ( err ) {
                log( rc + '\nsomething went wrong..\n' );
                log( rc + 'checksums:', gc + checksum, '!==', rc + resultChecksum, '\n' );                
                log( err.stack, ec, '\n' );
            }
        }, delay );
    } );

    filter.on( 'close', function () {
    } );

    // source stream

    sourceStream = fs.createReadStream( path, { bufferSize : 80 } );

    sourceStream.on( 'data', function ( data ) {
        log( gc + 'input data:', bc, data, ec );
        log( gc + 'input data length:', bc, data.length, ec )
    } );
    
    sourceStream.on( 'drain', function () {
    } );
        
    sourceStream.on( 'end', function () {
    } );
        
    sourceStream.on( 'close', function () {
    } );
        
    sourceStream.on( 'error', function ( err ) {
        log( rc + 'source stream emits error:', err.stack, ec + '\n' );
    } );
    

    var ofilter = new Dropper( 13 );
    
    ofilter.on( 'data', function ( data ) {
    } );

    sourceStream.pipe( ofilter, { end : true } );

    ofilter.pipe( filter );

} );




lib/dropper.js
/*!
 * dropper
 * Copyright(c) 2011 Guglielmo Ferri <44gatti@gmail.com>
 * MIT Licensed
 */

/**
 * Library version 0.0.6
 */

exports.version = '0.0.6';

var log = console.log,
    stream = require( 'stream' ).Stream,
    util = require( 'util' ),
    dropper = function ( dropSize, opt ) {
        var me = this,
            pad = opt && opt.pad;
        me.writable = true;
        me.readable = true;
        // super constructor call is not necessary here, no arguments
        me.constructor.super_.call( me );
        // public vars
        me.dropSize = dropSize || ( 64 * 1024 );
        me.pad = ( typeof pad === 'number' ) ? pad & 255 : false;
        me.sent = 0;
        me.received = 0;
        // private vars
        me._drop = null;
        me._cpos = 0;
    },
    dproto;

util.inherits( dropper, stream );

dproto = dropper.prototype;

dproto.pause = function () {
    var me = this;
    me.emit( 'pause' );
};

dproto.resume = function () {
    var me = this;
    me.emit( 'resume' );
};

dproto.destroy = function () {
    var me = this;
    me.emit( 'close' );
};

dproto.pipe = function ( dest, options ) {
    var me = this,
        d = dproto.__proto__.pipe.apply( me, arguments );
    // nodeJS 0.4.x Stream.pipe() doesn't return 
    // the destination stream, manually added the code 
    return ( d || dest );
};

dproto.write = function ( data ) {
    var me = this;
    ++me.received;
    if ( data.length > me.dropSize ) {
        // packet size > dropSize;
        me._chargeAndSqueeze( data );
    } else {
        // packet size <= dropSize;
        me._squeeze( data );
    }
    return true;
};

dproto._chargeAndSqueeze = function ( data ) {
    var me = this,
        dropSize = me.dropSize,
        dlen = data.length,
        offset = 0;
    if ( me._cpos ) {
        // dropper is not empty
        offset = dropSize - me._cpos;
        data.copy( me._drop, me._cpos, 0, offset );
        // a drop falls
        ++me.sent;
        me.emit( 'data', me._drop );
    }
    // fill up, then squeeze the dropper
    var i = offset,
        j = i + dropSize;
    for ( ; ; i = j, j += dropSize ) {
        if ( j > dlen - 1 ) {
              me._cpos = dlen - i;
              if ( j === dlen ) {
                  me._cpos = 0;
                  // a drop falls
                  ++me.sent;
                  me.emit( 'data', data.slice( i, j ) );
              } else {
                  me._drop = new Buffer( dropSize );
                  data.copy( me._drop, 0, i, dlen );
              }
              break;
          }
          me._cpos = 0;
          // a drop falls
          ++me.sent;
          me.emit( 'data', data.slice( i, j ) );
      }
};

dproto._squeeze = function ( data ) {
    var me = this,
        dropSize = me.dropSize,
        dlen = data.length,
        offset = 0;

    if ( me._cpos ) {
        // dropper is not empty
        offset = dropSize - me._cpos;
        if ( offset >= dlen ) {
            data.copy( me._drop, me._cpos, 0, dlen );
            me._cpos += dlen;
            if ( me._cpos === dropSize ) {
                me._cpos = 0;
                // a drop falls
                ++me.sent;
                me.emit( 'data', me._drop );
            }
        } else {
            data.copy( me._drop, me._cpos, 0, offset );
            // a drop falls
            ++me.sent;
            me.emit( 'data', me._drop );
            
            me._drop = new Buffer( dropSize );
            data.copy( me._drop, 0, offset, dlen );
            me._cpos += dlen - dropSize;
        }
    } else {
        // dropper is empty
        me._cpos = dlen;
        if ( me._cpos === dropSize ) {
            me._cpos = 0;
            // a drop falls
            ++me.sent;
            me.emit( 'data', data );
        } else {
            me._drop = new Buffer( dropSize );
            data.copy( me._drop, 0, 0, dlen );
        }
    }
};

dproto.end = function () {
    var me = this,
        ldrop = null,
        drop = me._drop,
        pos = me._cpos,
        pad = me.pad;
    // empty the dropper if contains something
    if ( pos ) {
        if ( pad || pad === 0 ) {
            var dropSize = me.dropSize;
                offset = dropSize - pos;
                i = 0;
            // create padding
            ldrop = new Buffer( dropSize );
            for ( ; i < offset ; ldrop[ i++ ] = pad );
            for ( offset = 0; i < dropSize; ldrop[ i++ ] = drop[ offset++ ] );
        } else {
            ldrop = drop.slice( 0, pos );
        }
        me._cpos = 0;
        me._drop = null;
        // the last drop falls
        ++me.sent;
        me.emit( 'data', ldrop );
    }
    me.emit( 'end' );
};

exports.dropper = dropper;


examples/simple.js
var log = console.log,
    fs = require( 'fs' ),
    Dropper = require( '../index' ),
    path = __dirname + '/numbers.txt',
    filter = null, rs = null,
    rc = '\033[1;31m', gc = '\033[1;32m', yc = '\033[1;33m', yc = '\033[1;33m', bc = '\033[1;39m', ec = '\033[0m';

fs.stat( path, function ( err, stats ) {
    if ( err ) {
        log( 'fs stat error:', err.stack );
    }
    log( rc + '\nsource file:', path + ', bytes:', stats.size, '\n' + ec );

} );

// dropper

filter = new Dropper( 36 );

filter.on( 'data', function ( data ) {
    filter.pause(); // <- pausing filter
    log( gc + 'filter emits data, length:', data.length + ec );
    log( bc + 'data:', data, ec, '\n' );
    filter.resume(); // <- resuming filter
} );

filter.on( 'drain', function () {
    log( 'filter emits drain\n' );
} );

filter.on( 'end', function () {
    log( rc + '\nfilter stream ends,', filter.sent, 'packet(s) sent,', filter.received, 'received\n' + ec );
} );

filter.on( 'close', function () {
    log( rc + 'filter stream was closed\n' + ec );
} );

// source stream

rs = fs.createReadStream( path, { bufferSize : 40 } );

rs.on( 'data', function ( data ) {
    log( yc + 'source stream emits data, length:', data.length + ec );
    log( bc + 'data:', data, ec );
    log();
} );
rs.on( 'drain', function () {
    log( 'source stream emits drain\n' );
} );
    
rs.on( 'end', function () {
    log( rc + 'source stream emits end\n' + ec );
} );
    
rs.on( 'close', function () {
    log( rc + 'source stream was closed\n' + ec );
} );
    
rs.on( 'error', function ( err ) {
    log( rc + 'source stream emits error:', err.stack, ec + '\n' );
} );

rs.pipe( filter, { end : true } );

